File: ts/project/.sdk/src/cmp/csharp/fragment/EntityBase.fragment.cs
// ProjectNameEntityBase - shared entity behaviour: construction, data/match
// state (with feature hooks), the operation pipeline (RunOp) and default
// unsupported-op implementations of every CRUD method. Generated entity
// classes derive from this and override the operations their API defines.

using System.Runtime.CompilerServices;

using Voxgig.Struct;

namespace ProjectNameSdk;

public abstract class ProjectNameEntityBase : IEntity
{
    protected string name;
    protected ProjectNameSDK client;
    protected Utility utility;
    protected Dictionary<string, object?> entopts;
    protected Dictionary<string, object?> data = new();
    protected Dictionary<string, object?> match = new();
    protected Context entctx;

    protected ProjectNameEntityBase(ProjectNameSDK client,
        Dictionary<string, object?>? entopts, string name)
    {
        entopts ??= new Dictionary<string, object?>();
        if (!entopts.ContainsKey("active"))
        {
            entopts["active"] = true;
        }
        else if (!Equals(entopts["active"], false))
        {
            entopts["active"] = true;
        }

        this.name = name;
        this.client = client;
        this.utility = client.GetUtility();
        this.entopts = entopts;

        this.entctx = utility.MakeContext(new Dictionary<string, object?>
        {
            ["entity"] = this,
            ["entopts"] = entopts,
        }, client.GetRootCtx());

        utility.FeatureHook(this.entctx, "PostConstructEntity");
    }

    public string GetName() => name;

    public abstract IEntity Make();

    protected Dictionary<string, object?> CloneOpts()
    {
        return new Dictionary<string, object?>(entopts);
    }

    public object? Data(object? newdata = null)
    {
        if (newdata != null)
        {
            data = Helpers.ToMapAny(StructUtils.Clone(newdata))
                ?? new Dictionary<string, object?>();
            utility.FeatureHook(entctx, "SetData");
        }

        utility.FeatureHook(entctx, "GetData");
        return StructUtils.Clone(data);
    }

    public object? Match(object? newmatch = null)
    {
        if (newmatch != null)
        {
            match = Helpers.ToMapAny(StructUtils.Clone(newmatch))
                ?? new Dictionary<string, object?>();
            utility.FeatureHook(entctx, "SetMatch");
        }

        utility.FeatureHook(entctx, "GetMatch");
        return StructUtils.Clone(match);
    }

    public virtual object? Load(Dictionary<string, object?>? reqmatch,
        Dictionary<string, object?>? ctrl = null)
        => throw Helpers.UnsupportedOp("load", name);

    public virtual object? List(Dictionary<string, object?>? reqmatch,
        Dictionary<string, object?>? ctrl = null)
        => throw Helpers.UnsupportedOp("list", name);

    public virtual object? Create(Dictionary<string, object?>? reqdata,
        Dictionary<string, object?>? ctrl = null)
        => throw Helpers.UnsupportedOp("create", name);

    public virtual object? Update(Dictionary<string, object?>? reqdata,
        Dictionary<string, object?>? ctrl = null)
        => throw Helpers.UnsupportedOp("update", name);

    public virtual object? Remove(Dictionary<string, object?>? reqmatch,
        Dictionary<string, object?>? ctrl = null)
        => throw Helpers.UnsupportedOp("remove", name);

    protected object? RunOp(Context ctx, Action postDone)
    {
        // #PrePoint-Hook

        try
        {
            var point = utility.MakePoint(ctx);
            ctx.Out["point"] = point;
        }
        catch (Exception err)
        {
            return utility.MakeError(ctx, err);
        }

        // #PreSpec-Hook

        try
        {
            var spec = utility.MakeSpec(ctx);
            ctx.Out["spec"] = spec;
        }
        catch (Exception err)
        {
            return utility.MakeError(ctx, err);
        }

        // #PreRequest-Hook

        try
        {
            var resp = utility.MakeRequest(ctx);
            ctx.Out["request"] = resp;
        }
        catch (Exception err)
        {
            return utility.MakeError(ctx, err);
        }

        // #PreResponse-Hook

        try
        {
            var resp2 = utility.MakeResponse(ctx);
            ctx.Out["response"] = resp2;
        }
        catch (Exception err)
        {
            return utility.MakeError(ctx, err);
        }

        // #PreResult-Hook

        try
        {
            var result = utility.MakeResult(ctx);
            ctx.Out["result"] = result;
        }
        catch (Exception err)
        {
            return utility.MakeError(ctx, err);
        }

        // #PreDone-Hook

        postDone();

        return utility.Done(ctx);
    }

    // Streaming operations. Runs `action` through the full pipeline and returns
    // an async iterator over result items, so the `streaming` feature's
    // incremental output is reachable from a generated entity (a normal op call
    // materialises the whole result). `callopts` parameterises the call:
    //   - inbound (download): iterate the yielded items/chunks (from the
    //     streaming feature when active, else the materialised items);
    //   - outbound (upload): pass a streamable payload as callopts["body"] - it
    //     is attached to the request so the transport can send it;
    //   - callopts["ctrl"] threads pipeline control and callopts["signal"] (a
    //     CancellationToken) is honoured between yields.
    public async IAsyncEnumerable<object?> Stream(
        string action,
        Dictionary<string, object?>? args = null,
        Dictionary<string, object?>? callopts = null,
        [EnumeratorCancellation] CancellationToken cancel = default)
    {
        callopts ??= new Dictionary<string, object?>();

        var signal =
            StructUtils.GetProp(callopts, "signal") is CancellationToken sigTok
                ? sigTok
                : CancellationToken.None;

        var ctrl = Helpers.ToMapAny(StructUtils.GetProp(callopts, "ctrl"))
            ?? new Dictionary<string, object?>();
        ctrl["stream"] = callopts;

        var ctxmap = new Dictionary<string, object?>
        {
            ["opname"] = action,
            ["ctrl"] = ctrl,
            ["match"] = match,
            ["data"] = data,
        };
        if (args != null)
        {
            foreach (var kv in args)
            {
                ctxmap[kv.Key] = kv.Value;
            }
        }

        var ctx = utility.MakeContext(ctxmap, entctx);

        // Outbound: expose the caller's streamable payload so the request
        // builder / transport can stream it as the request body.
        var body = StructUtils.GetProp(callopts, "body");
        if (body != null)
        {
            ctx.Reqdata["body$"] = body;
            ctx.Ctrl.Stream_out = body;
        }

        // Run the same pipeline the op methods run.
        var materialised = RunOp(ctx, () =>
        {
            if (ctx.Result?.Resmatch != null)
            {
                match = ctx.Result.Resmatch;
            }
        });

        await Task.CompletedTask;

        // Inbound: prefer the streaming feature's incremental iterator; else
        // fall back to the materialised items so `stream` always yields.
        var stream = ctx.Result?.Stream;
        if (stream != null)
        {
            foreach (var item in stream())
            {
                if (cancel.IsCancellationRequested || signal.IsCancellationRequested)
                {
                    yield break;
                }
                yield return item;
            }
        }
        else
        {
            var items = materialised is List<object?> list
                ? list
                : (materialised == null
                    ? new List<object?>()
                    : new List<object?> { materialised });
            foreach (var item in items)
            {
                if (cancel.IsCancellationRequested || signal.IsCancellationRequested)
                {
                    yield break;
                }
                yield return item;
            }
        }
    }
}

File: ts/project/.sdk/src/cmp/swift/fragment/EntityBase.fragment.swift
// ProjectNameEntityBase - shared entity behaviour: construction, data/match
// state (with feature hooks), the operation pipeline (runOp) and default
// unsupported-op implementations of every CRUD method. Generated entity
// classes derive from this and override the operations their API defines.

import Foundation

open class ProjectNameEntityBase: Entity {
  var name: String
  var client: ProjectNameSDK
  var utility: Utility
  var entopts: VMap
  var data = VMap()
  var match = VMap()
  var entctx: Context!

  public init(_ client: ProjectNameSDK, _ entoptsIn: VMap?, _ name: String) {
    let entopts = entoptsIn ?? VMap()
    if entopts.entries["active"] == nil {
      entopts.entries["active"] = .bool(true)
    } else if entopts.entries["active"] != .bool(false) {
      entopts.entries["active"] = .bool(true)
    }

    self.name = name
    self.client = client
    self.utility = client.getUtility()
    self.entopts = entopts

    self.entctx = utility.makeContext(["entity": self, "entopts": entopts], client.getRootCtx())

    utility.featureHook(entctx, "PostConstructEntity")
  }

  public func getName() -> String { name }

  open func make() -> Entity {
    return ProjectNameEntityBase(client, cloneOpts(), name)
  }

  func cloneOpts() -> VMap {
    let m = VMap()
    for (k, v) in entopts.entries { m.entries[k] = v }
    return m
  }

  @discardableResult
  public func data(_ newdata: Value?) -> Value {
    if let nd = newdata, !isNil(nd) {
      data = clone(nd).asMap ?? VMap()
      utility.featureHook(entctx, "SetData")
    }
    utility.featureHook(entctx, "GetData")
    return clone(.map(data))
  }

  @discardableResult
  public func matchv(_ newmatch: Value?) -> Value {
    if let nm = newmatch, !isNil(nm) {
      match = clone(nm).asMap ?? VMap()
      utility.featureHook(entctx, "SetMatch")
    }
    utility.featureHook(entctx, "GetMatch")
    return clone(.map(match))
  }

  open func load(_ reqmatch: VMap?, _ ctrl: VMap?) throws -> Value {
    throw unsupportedOp("load", name)
  }

  open func list(_ reqmatch: VMap?, _ ctrl: VMap?) throws -> Value {
    throw unsupportedOp("list", name)
  }

  open func create(_ reqdata: VMap?, _ ctrl: VMap?) throws -> Value {
    throw unsupportedOp("create", name)
  }

  open func update(_ reqdata: VMap?, _ ctrl: VMap?) throws -> Value {
    throw unsupportedOp("update", name)
  }

  open func remove(_ reqmatch: VMap?, _ ctrl: VMap?) throws -> Value {
    throw unsupportedOp("remove", name)
  }

  func runOp(_ ctx: Context, _ postDone: () -> Void) throws -> Value {
    // #PrePoint-Hook

    do {
      let point = try utility.makePoint(ctx)
      ctx.out["point"] = point
    } catch {
      return try utility.makeError(ctx, error)
    }

    // #PreSpec-Hook

    do {
      let spec = try utility.makeSpec(ctx)
      ctx.out["spec"] = spec
    } catch {
      return try utility.makeError(ctx, error)
    }

    // #PreRequest-Hook

    do {
      let resp = try utility.makeRequest(ctx)
      ctx.out["request"] = resp
    } catch {
      return try utility.makeError(ctx, error)
    }

    // #PreResponse-Hook

    do {
      let resp2 = try utility.makeResponse(ctx)
      ctx.out["response"] = resp2
    } catch {
      return try utility.makeError(ctx, error)
    }

    // #PreResult-Hook

    do {
      let result = try utility.makeResult(ctx)
      ctx.out["result"] = result
    } catch {
      return try utility.makeError(ctx, error)
    }

    // #PreDone-Hook

    postDone()

    return try utility.done(ctx)
  }

  // Streaming operations. Runs `action` through the full pipeline and returns
  // an AsyncStream over result items, so the `streaming` feature's incremental
  // output is reachable from a generated entity (a normal op call materialises
  // the whole result). `callopts` parameterises the call:
  //   - inbound (download): iterate the yielded items/chunks (from the
  //     streaming feature when active, else the materialised items);
  //   - outbound (upload): pass a streamable payload as callopts["body"] - it
  //     is attached to the request so the transport can send it;
  //   - callopts["ctrl"] threads pipeline control and callopts["signal"] (a
  //     native () -> Bool that returns true when aborted) is honoured between
  //     yields.
  public func stream(_ action: String, _ args: VMap? = nil, _ callopts: VMap? = nil)
    throws -> AsyncStream<Value>
  {
    let opts = callopts ?? VMap()

    let signal = opts.entries["signal"]?.asNative as? @Sendable () -> Bool

    let ctrl = opts.entries["ctrl"]?.asMap ?? VMap()
    ctrl.entries["stream"] = .map(opts)

    var ctxmap: [String: Any?] = [
      "opname": action,
      "ctrl": ctrl,
      "match": match,
      "data": data,
    ]
    if let a = args {
      for (k, v) in a.entries { ctxmap[k] = v }
    }

    let ctx = utility.makeContext(ctxmap, entctx)

    // Outbound: expose the caller's streamable payload so the request builder
    // / transport can stream it as the request body.
    if let body = opts.entries["body"], !isNil(body) {
      ctx.reqdata.entries["body$"] = body
      ctx.ctrl.streamOut = body
    }

    // Run the same pipeline the op methods run.
    let materialised = try runOp(ctx, {})

    // Inbound: prefer the streaming feature's incremental iterator; else fall
    // back to the materialised items so `stream` always yields.
    var items: [Value] = []
    if let result = ctx.result, let streamFn = result.stream {
      items = streamFn()
    } else if let list = materialised.asList {
      items = list.items
    } else if !isNil(materialised) {
      items = [materialised]
    }

    return AsyncStream { continuation in
      for item in items {
        if let s = signal, s() { break }
        continuation.yield(item)
      }
      continuation.finish()
    }
  }
}
